Support precomputed_embeddings for Llama 4

AlienKevin opened the pull request:
## Motivation

Support precomputed_embeddings for llama 4 to close [#8065](https://github.com/sgl-project/sglang/issues/8065)

## Modifications

1. Simplified mllama4 processor by calling `process_and_combine_mm_data`.
2. Added an example notebook for running vision queries with llama 4, including using precomputed_embeddings

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [x] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

Commit:
Support precomputed_features for Llama 4

JustinTong0323 commented on python/sglang/srt/layers/multimodal.py:
```diff
@@ -55,14 +55,17 @@ def gpu_tensor_hash(tensor: torch.Tensor) -> int:
-        XCONST=PRIME_2,
-    )
+    # Set cuda device to prevent ValueError: Pointer argument (at 0) cannot be accessed from Triton (cpu tensor?)
+    # Solution from Tri: https://github.com/Dao-AILab/flash-attention/issues/523#issuecomment-1707611579
```
Thanks for solving that!

JustinTong0323 commented on python/sglang/srt/multimodal/processors/mllama4.py:
Have you tested it? I tried this before but the model could not recognize images and output wrong answer.

JustinTong0323 commented on docs/backend/vlm_query_llama4.ipynb:
maybe add to "sglang/docs/backend/vlm_query.ipynb"

JustinTong0323 reviewed (commented):
Thanks for the contribution!

Commit:
Merge vlm_query_llama4.ipynb into vlm_query.ipynb

AlienKevin commented:
@JustinTong0323 Tested with meta-llama/Llama-4-Maverick-17B-128E-Instruct-FP8 on the sample image captioning task in the notebook. The captioning is consistent with the output of the raw image version. Also merged the llama 4 parts into the existing notebook so you can check the caption outputs in that notebook.

JustinTong0323 committed:
Merge branch 'main' into llama4_precomputed_features

Commit:
Rename precomputed_features to precomputed_embeddings

Commit:
Added TestLlama4ImageUnderstandsImage to test_vlm_input_format.py

AlienKevin commented:
@JustinTong0323 Renamed precomputed_features to precomputed_embeddings and added llama4 tests to test_vlm_input_format.py (all 3 tests passed on Maverick).

AlienKevin commented:
@JustinTong0323 Let me know if there's any additional things we need before the merge

JustinTong0323 committed:
Merge branch 'main' into llama4_precomputed_features

JustinTong0323 commented:
> @JustinTong0323 Let me know if there's any additional things we need before the merge

I'm really sorry for the late response. The CI failed because the llama4 is too large for the CI machines. Perhaps we could comment out the llama4 test, run it only locally, and then paste the results right here.
Also, it would be better to contact me on slack for quicker reply. (as there are too many github msgs))

JustinTong0323 committed:
update: Add VLM query notebook and update model path in tests

- Added `vlm_query.ipynb` to the documentation, enhancing the backend resources.
- Updated the model path in `test_vlm_input_format.py` to use `Llama-4-Scout-17B-16E-Instruct` instead of `Llama-4-Maverick-17B-128E-Instruct-FP8`.
- Adjusted execution counts and outputs in the notebook for better clarity and functionality.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
not run llama4 in ci

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
update

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 commented:
Thanks for the contribution, great work!

JustinTong0323 committed:
Merge branch 'main' into llama4_precomputed_features

JustinTong0323 committed:
Update pre-commit configuration to exclude additional files from codespell checks

- Added `docs/backend/vlm_query.ipynb` to the exclusion list alongside `test/srt/test_reasoning_parser.py` to prevent false positives during linting.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
Merge branch 'main' into llama4_precomputed_features

JustinTong0323 committed:
fix: donot run llama4 in CI

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>
